### ISSUE TITLE
@JsonFormat Support for locale specific number format

### ISSUE THREAD

**Issue by MBurchard** [FasterXML/jackson-databind#2945]
**Describe the bug**
@JsonFormat does not work as expected when converting string to double and vice versa.

**Version information**
2.11.3

**To Reproduce**
I have an entity with a double property. Challenge is to read an external API that sends the capacity value as string in German Format.
```kotlin
@JsonIgnoreProperties(ignoreUnknown = true)
data class Event(
		@JsonProperty("NUMCapacity")
		@JsonFormat(shape = Shape.STRING, locale = "de_DE", pattern = "#.#")
		var capacity: Double,
```

```
Caused by: org.springframework.http.converter.HttpMessageNotReadableException: JSON parse error: Cannot deserialize value of type `double` from String "158,1": not a valid Double value; nested exception is com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `double` from String "158,1": not a valid Double value
```

**Expected behavior**
Expected behaviour is, when given the German locale, the comma is understood as the decimal separator.

**Comment by cowtowncoder** [comment]
Makes sense that this could be used; as you have noticed, that neither explicit `pattern` not Locale-dependant default pattern is supported for "numbers-as-Strings" coercion case currently.

Should be support at least for float/Float, double/Double, BigDecimal (?). And probably integer counterparts too.

**Comment by Gaibhne** [comment]
Is this issue abandoned ? I wanted to use Jackson for parsing numbers into BigDecimals too and stumbled over this same problem.

**Comment by MBurchard** [comment]
> Is this issue abandoned ? I wanted to use Jackson for parsing numbers into BigDecimals too and stumbled over this same problem.

As you see, nothing happens in more than a year. So It has at least no priority.

**Comment by cowtowncoder** [comment]
In OSS projects it all comes down to someone having time and itch to work on things -- I don't think anyone has worked on this, no.

Contributions welcome: this looks like something that would be relatively easy to make progress on. Will add a label to suggest that.

**Comment by emiatej9** [comment]
Searching code related to this issue, I found below lines in `deserializeAndSet` method:

https://github.com/FasterXML/jackson-databind/blob/ada34dcf6676aa5c5869791c23143a65a7bf56c2/src/main/java/com/fasterxml/jackson/databind/deser/impl/FieldProperty.java#L135-L140

In line 138 `_valueDeserializer` have reference of static class `DoubleDeserializer` in `NumberDeserializers`. I think the annotations the `_field` property is holding should be passed to `DoubleDeserializer#deserialize(p, ctxt)` at this point so as to exploit `locale` and `pattern` before parsing the given text.

https://github.com/FasterXML/jackson-databind/blob/ada34dcf6676aa5c5869791c23143a65a7bf56c2/src/main/java/com/fasterxml/jackson/databind/deser/impl/FieldProperty.java#L31-L36

However, It seems very difficult to modify parameters of `deserialize(Jsonparser p, DeserializationContext ctxt)` method which is inherited from `JsonDeserialize` class.

Could you give me some hints or guides to fix this issue?

**Comment by cowtowncoder** [comment]
@emiatej9 Typically all handling of format annotations should be done during construction or contextualization phases and specifically should never be passed (and rarely if ever used) accessed during actual deserialization calls.

Contextualization refers to handling in method `createContextual()`, which is called after construction but before any actual use for deserialization. It will be passed `BeanProperty` through which annotations (property and class annotations) are accessible.

You don't need to check `FieldProperty` methods: this is not the place handling should occur (since these are deserializer-independent entities).

Instead, `createContextual()` method -- you may want to see how other standard deserializers use it -- usually checks for various configuration settings, annotations, and so on, and then creates (and returns) a differently configured instance as necessary.

**Comment by emiatej9** [comment]
@cowtowncoder Thank you for your help.

I examined contextualization process as you mentioned, and I found `DateBasedDeserializer` also has similar format issue.

If numerical deserializers including `DoubleDeserializer` could implement  `ContextualDeserializer`
as`DateBasedDeserializer` did, I can handle formatting issue in `createContextual` method by delegating deserializing to a brand-new deserializer like `FormatAnnotatedNumberDeserializer`.

```java
@JacksonStdImpl
public static class DoubleDeserializer 
    extends PrimitiveOrWrapperDeserializer<Double>
    implements ContextualDeserializer
{

    ...

    @Override
    public JsonDeserializer<?> createContextual(DeserializationContext ctxt, BeanProperty property) throws JsonMappingException {
        final JsonFormat.Value format = findFormatOverrides(ctxt, property, handledType());
        if (format != null) {
            return new FormatAnnotatedNumberDeserializer(this, format);
        }
        return this;
    }
    ...
}

```
or introducing delegate property, and then use it in `deserialize(p, ctxt)`:

```java
@JacksonStdImpl
public static class DoubleDeserializer 
    extends PrimitiveOrWrapperDeserializer<Double>
    implements ContextualDeserializer
{
    ...
    
    protected JsonDeserializer<Number> delegatee = null;

    @Override
    public JsonDeserializer<?> createContextual(DeserializationContext ctxt, BeanProperty property) throws JsonMappingException {
        final JsonFormat.Value format = findFormatOverrides(ctxt, property, handledType());
        if (format != null) {
            this.delegatee = new FormatAnnotatedNumberDeserializer(format);
        }
        return this;
    }

    @Override
    public Double deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {
        if (delegatee != null)  {
            Double value = delegatee.deserialize(p, ctxt).doubleValue();             
            if (value != null) {
                return value;
            }
        }
        ...
        // if delegatee is null, deserialize as usual
    }
    ...
}


```

What do you think of this idea?

**Comment by emiatej9** [comment]
I also found another issue in parsing locale string like 'de_DE'.

In `JsonFormat.java`, It seems `localeStr` couldn't be parsed with `new Locale(localeStr)`:

```java
public Value(String p, Shape sh, String localeStr, String tzStr, Features f, Boolean lenient)
{
    this(p, sh, 
        (localeStr == null || localeStr.length() == 0 || DEFAULT_LOCALE.equals(localeStr)) ? null : new Locale(localeStr),
        (tzStr == null || tzStr.length() == 0 || DEFAULT_TIMEZONE.equals(tzStr)) ? null : tzStr, 
        null, f, lenient);
}
```

**Comment by krish2kdev** [comment]
@cowtowncoder 
can i work on this issue?

**Comment by JooHyukKim** [comment]
Read this issue and PR #3485 thoroughly and you will find out probability of resolving this issue is very low.

**Comment by JooHyukKim** [comment]
Read this issue and PR https://github.com/FasterXML/jackson-databind/pull/3485 thoroughly and you will find out probability of resolving this issue is very low @krish2kdev

**Comment by cowtowncoder** [comment]
Yeah... I am not sure if I wanted such feature or not. Seems like something for which custom serializer/deserializer combination might make most sense.
Things just get quite complicated.
